hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java

```diff
@@ -17,92 +17,120 @@
+ * methods have been created for this purpose.
+ *
+ * <p>{@link #requireKey(HederaKey)} is used to add a required non-payer signing key (remember, the payer signing
+ * key was added when the context was created). Some basic validation is performed (the key cannot be null or empty).
  */
-public class PreHandleContext {
-
+public final class PreHandleContext {
+    /** Used to get keys for accounts and contracts. */
     private final AccountAccess accountAccess;
-
+    /** The transaction body. */
     private final TransactionBody txn;
+    /** The payer account ID. Specified in the transaction body, extracted and stored separately for convenience. */
     private final AccountID payer;
-    private final List<HederaKey> requiredNonPayerKeys = new ArrayList<>();
-
-    private ResponseCodeEnum status;
-    private HederaKey payerKey;
+    /** The payer's key, as found in state */
+    private final HederaKey payerKey;
+    /**
+     * The set of all required non-payer keys. A {@link LinkedHashSet} is used to maintain a consistent ordering.
+     * While not strictly necessary, it is useful at the moment to ensure tests are deterministic. The tests should
+     * be updated to compare set contents rather than ordering.
+     */
+    private final Set<HederaKey> requiredNonPayerKeys = new LinkedHashSet<>();
+    /** Scheduled transactions have a secondary "inner context". Seems not quite right. */
     private PreHandleContext innerContext;
 
-    public PreHandleContext(@NonNull final AccountAccess accountAccess, @NonNull final TransactionBody txn) {
-        this(accountAccess, txn, txn.transactionIDOrElse(TransactionID.DEFAULT).accountIDOrElse(AccountID.DEFAULT), OK);
-    }
-
-    public PreHandleContext(
-            @NonNull final AccountAccess accountAccess,
-            @NonNull final TransactionBody txn,
-            @NonNull final AccountID payer) {
-        this(accountAccess, txn, payer, OK);
-    }
-
-    public PreHandleContext(
-            @NonNull final AccountAccess accountAccess,
-            @NonNull final TransactionBody txn,
-            @NonNull final ResponseCodeEnum status) {
+    /**
+     * Create a new PreHandleContext instance. The payer and key will be extracted from the transaction body.
+     *
+     * @param accountAccess used to get keys for accounts and contracts
+     * @param txn the transaction body
+     * @throws PreCheckException if the payer account ID is invalid or the key is null
+     */
+    public PreHandleContext(@NonNull final AccountAccess accountAccess, @NonNull final TransactionBody txn)
+            throws PreCheckException {
         this(
                 accountAccess,
                 txn,
                 txn.transactionIDOrElse(TransactionID.DEFAULT).accountIDOrElse(AccountID.DEFAULT),
-                status);
+                INVALID_PAYER_ACCOUNT_ID);
     }
 
-    public PreHandleContext(
+    /** Create a new instance */
+    private PreHandleContext(
```

Comment:
The checks in this constructor are specific, so it's highly unlikely we would ever correctly call this private constructor with a different response code enum.  Would it not be simpler to merge this logic into the public constructor (and set the specific response code at the point the check is made) without the extra indirection?

Reply:
The problem is in the "createInnerContext" method, which also calls this constructor but uses a user supplied value. Initially I just hard coded the error as you suggested but when I got to the inner context I realized I needed to make this an arg

Reply:
hmmm, not sure about the inner context item; probably needs a closer look to be sure we're doing the right things there.
Not an issue for this PR, however.

Reply:
Agree, the inner context bothers me because a special need for Scheduled service found its way into this common API, which seems like something is wrong. But we can address it in another PR.